### PR TITLE
[WIP] Organization SMS Twilio Feature Updates Roadmap

### LINKED ISSUE #4815: Add SMS to Organization Features Section
**What type(s) of user does this feature affect?**
 - admins

**Description**
On the edit casa org page,
 - Create a new checkbox for SMS under the organization features section
 - Move the twilio fields under the new checkbox.
 - Only show the twilio text fields if the SMS checkbox is checked
 - Is the SMS checkbox is unchecked, disable all SMS features
 - Preserve the twilio field values when the SMS checkbox is unchecked

**Screenshots of current behavior, if any**  
image

_Login Details:_  
[Link to QA site](https://casa-qa.herokuapp.com/)  

Login Emails: 
- volunteer1@example.com  view site as a volunteer
- supervisor1@example.com view site as a supervisor
- casa_admin1@example.com view site as an admin
- all_casa_admin1@example.com view site as an all casa admin
  - go to `/all_casa_admins/sign_in`  

password for all users: 12345678  

### Questions? Join Slack!

We highly recommend that you join us in slack https://rubyforgood.herokuapp.com/ #casa channel to ask questions quickly and hear about office hours (currently Tuesday 5-7pm Pacific), stakeholder news, and upcoming new issues.

### PR DESCRIPTION
**Pull request by ShamiTomita**
### What github issue is this PR for, if any?
Resolves #4815

### What changed, and why?
Changes to Twilio Functionality in frontend and backend. 
On Edit Casa Organization, the Twilio Form Inputs have been nested in an accordion drop down within the Organization Features section. Currently, if the Twilio Enabled checkbox is checked (and if the rest of the form has valid inputs), the CASA org will validate its Twilio Credentials upon update. If the Twilio checkbox is disabled, any existing Twilio credentials are still stored but any SMS services should return a notice saying: "SMS notice was not sent. Error: Twilio Is Not Enabled."

The TwilioService now has its own custom error that handles checking to see if a casa_org has twilio enabled. 

Changes include: 
 - [x] adding a twilio_enabled boolean column to the casa_org table 
 - [x] creating a casa_org.js file to handle Twilio form state (adding or removing "required" or "disabled" depending on Twilio Enabled Checkbox)
 - [x] sms_reactivation_toggle.js to prevent SMS activity on the frontend if twilio is disabled

### How will this affect user permissions?
- Volunteer permissions: N/A
- Supervisor permissions: N/A
- Admin permissions: Changes to Twilio form in EDIT Casa Org

### How is this tested? (please write tests!) 💖💪
Relevant tests added for all three user types, application controller, requests/users/password_spec.rb

### Screenshots please :)
Screenshot 2023-05-22 at 10 19 58 AM
Screenshot 2023-05-22 at 10 20 25 AM
Screenshot 2023-05-31 at 4 15 08 PM
Screenshot 2023-05-31 at 4 15 18 PM

### Feelings gif (optional)
What gif best describes your feeling working on this issue? https://giphy.com/
How to embed:
`alt text`

### Feedback please? (optional)
We are very interested in your feedback! Please give us some :) https://forms.gle/1D5ACNgTs2u9gSdh9

### CODE REVIEW

**Comment by ShamiTomita** [comment]
Hi @cnorm35, if you have some time would you be able to give this a quick glance? Thank you!

File: Gemfile.lock [hunk 1]
```diff
@@ -1,79 +1,79 @@
 GEM
   remote: https://rubygems.org/
   specs:
-    actioncable (7.0.4.3)
-      actionpack (= 7.0.4.3)
-      activesupport (= 7.0.4.3)
+    actioncable (7.0.5)
```
Comment (compwron) [inline Gemfile.lock:4]: Please upgrade rails in a separate PR, this is a lot for one PR! :)

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -49,36 +49,40 @@ def deliver_sms_to(resource, body_msg)
...
       twilio_res.error_code.nil? ? "sent" : "error"
-    rescue Twilio::REST::RestError
+    rescue TwilioService::TwilioCasaOrgError => error
+      @error = error
+      "error"
+    rescue Twilio::REST::RestError => error
+      @error = error
+      "error" # Find a better description for this
+    rescue # unverfied error isnt picked up by Twilio::Rest::RestError
+      # https://www.twilio.com/docs/errors/21608
```
Comment (compwron) [inline app/controllers/application_controller.rb:73]: good comments :)

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -49,36 +49,40 @@ def deliver_sms_to(resource, body_msg)
...
 
-  def sms_acct_creation_notice(resource_name, sms_status)
-    if sms_status === "blank"
-      return "New #{resource_name} created successfully."
-    end
-    if sms_status === "error"
-      return "New #{resource_name} created successfully. SMS not sent due to error."
-    end
-    if sms_status === "sent"
+  def sms_acct_creation_notice(resource_name, sms_status) # switch case implement here (should read mutually exclusive)
```
Comment (compwron) [inline app/controllers/application_controller.rb:79]: seems reasonable. The comment seems excessive to me but I don't have a strong opinion.

File: app/services/twilio_service.rb [hunk 4]
```diff
@@ -26,8 +30,9 @@ def send_sms(params)
         body: body,
         to: to
       )
-    rescue => e
-      Rails.logger.error("send SMS failed: #{e}")
+    rescue => error
+      Rails.logger.error("send SMS failed: #{error}")
```
Comment (compwron) [inline app/services/twilio_service.rb:34]: I would rather see a bugsnag about it but I don't know that we're set up for that...

**Review by compwron — COMMENTED** [review]
Seems good :)

File: app/controllers/application_controller.rb (hunk 3 above)
Comment (ShamiTomita) [reply app/controllers/application_controller.rb:79]: Oh my bad that comment should've been scrubbed. It's a note from a pairing session.

File: Gemfile.lock (hunk 1 above)
Comment (ShamiTomita) [reply Gemfile.lock:4]: Sorry I can remove that from this PR.

**Comment by ShamiTomita** [comment]
I'm sorry for all of the commits, I was trying to squash all the small ones but ended up adding more >.< . I think all of the files are fine now though.

**Comment by cnorm35** [comment]
> Hi @cnorm35, if you have some time would you be able to give this a quick glance? Thank you!

Great job getting all of this together @ShamiTomita!

**Pull request closed without merging**